optimizely/event/dispatcher.go

```diff
@@ -54,10 +55,70 @@ func (*HTTPEventDispatcher) DispatchEvent(event LogEvent, callback func(bool)) {
 		if resp.StatusCode == 204 {
 			success = true
 		} else {
-			fmt.Printf("http.Post invalid response %d", resp.StatusCode)
+			dispatcherLogger.Error(fmt.Sprintf("http.Post invalid response %d", resp.StatusCode), nil)
 			success = false
 		}
 	}
 	callback(success)
+}
+
+type QueueEventDispatcher struct {
+	eventQueue Queue
+	mux        sync.Mutex
+	dispatcher *HTTPEventDispatcher
+}
+
+func (ed *QueueEventDispatcher) DispatchEvent(event LogEvent, callback func(bool)) {
+
+	ed.eventQueue.Add(event)
+	callback(true)
+	go func() {
+		ed.flushEvents()
+	}()
+}
+
+func (ed *QueueEventDispatcher) flushEvents() {
+
+	ed.mux.Lock()
+
+	defer func(){
+		ed.mux.Unlock()
+	}()
+
+	for ed.eventQueue.Size() > 0 {
+		items := ed.eventQueue.Get(1)
+		if len(items) == 0 {
+			time.Sleep(5 * time.Second)
+			continue
+		}
+		event, ok := items[0].(LogEvent)
+		if !ok {
+			//remove it?
+			dispatcherLogger.Error("invalid type passed to event dispatcher", nil)
+			ed.eventQueue.Remove(1)
+			continue
+		}
+
+		ed.dispatcher.DispatchEvent(event, func(success bool) {
+			if success {
+				ed.eventQueue.Remove(1)
+			} else {
+				// we failed.  Sleep 5 seconds and try again.  Or, should we exit and try the next time
```

Comment:
So if another log event is added, do they both get flushed? It looks like we only flush 1 at a time

Reply:
no, they are sent one at a time but all are flushed.  as in for queue.size > 0